components/remote_settings/src/client.rs

```diff
@@ -0,0 +1,551 @@
+    /// bucket, and collection defined by the [ClientConfig] used to generate
+    /// this [Client]. This function will return the raw network [Response].
+    pub fn get_records_raw(&self) -> Result<Response> {
+        let path = format!(
+            "v1/buckets/{}/collections/{}/records",
+            &self.bucket_name, &self.collection_name
+        );
+        self.make_request(path)
+    }
+
+    /// Fetches all records that have been published since provided timestamp
+    /// for a collection that can be found in the server, bucket, and
+    /// collection defined by the [ClientConfig] used to generate this [Client].
+    pub fn get_records_since(&self, timestamp: u64) -> Result<RemoteSettingsResponse> {
+        let path = format!(
+            "v1/buckets/{}/collections/{}/records?_since={}",
+            &self.bucket_name, &self.collection_name, timestamp
+        );
+        let resp = self.make_request(path)?;
+        let records = resp.json::<RecordsResponse>()?.data;
+        let last_modified = resp.headers.get_as("etag").unwrap()?;
+        Ok(RemoteSettingsResponse {
+            records,
+            last_modified,
+        })
+    }
+
+    /// Downloads an attachment from [attachment_location]. NOTE: there are no
+    /// guarantees about a maximum size, so use care when fetching potentially
+    /// large attachments.
+    pub fn get_attachment(&self, attachment_location: &str) -> Result<Response> {
+        let mut current_remote_state = self.remote_state.lock();
+        self.ensure_no_backoff(&mut current_remote_state.backoff)?;
+        let attachments_base_url = match &current_remote_state.attachments_base_url {
+            Some(url) => url.clone(),
+            None => {
+                let req = Request::get(self.base_url.clone());
+                let server_info = req.send()?.json::<ServerInfo>()?;
+                match server_info.capabilities.attachments {
+                    Some(capability) => Url::parse(&capability.base_url)?,
+                    None => Err(RemoteSettingsError::AttachmentsUnsupportedError)?,
+                }
+            }
+        };
+        current_remote_state.attachments_base_url = Some(attachments_base_url.clone());
+        drop(current_remote_state);
+
+        self.make_request(attachments_base_url.join(attachment_location)?.into())
+    }
+
+    fn make_request(&self, path: String) -> Result<Response> {
+        let mut current_remote_state = self.remote_state.lock();
+        self.ensure_no_backoff(&mut current_remote_state.backoff)?;
+        drop(current_remote_state);
+
+        let url = self.base_url.join(&path)?;
+        let req = Request::get(url);
+        let resp = req.send()?;
+
+        let mut current_remote_state = self.remote_state.lock();
+        self.handle_backoff_hint(&resp, &mut current_remote_state.backoff)?;
+
+        if resp.is_success() || resp.status == status_codes::NOT_MODIFIED {
+            Ok(resp)
```

Comment:
If the body really is empty here, a possibility might be a new enum, or even `Result<Option<Response>>`?

Reply:
This relates to the 304 handling above, correct?

Reply:
Correct, let's just drop the 304 handling, and then we don't have to worry about empty bodies.